### PR TITLE
big refactoring of Trainer

### LINKED ISSUE #204: Decompose `Trainer.train()` into smaller methods
It's grown to be a ~150 line method that's pretty hard to reason about.  It's a bit tricky to decompose, though, because of all of the dependencies between different parts of the method, but we should be able to pull out a bunch of it into separate methods.

### PR DESCRIPTION
**Pull request by joelgrus**
* broke `Trainer.train()` up into a lot of different functions
* got rid of all the `None` checks by writing wrapper functions that no-op in those cases
* created `TensorboardWriter` abstraction for the same reason

let me know if you don't like any of the design choices here

Fixes #204

### CODE REVIEW

File: allennlp/training/trainer.py [hunk 1]
```diff
@@ -125,18 +143,16 @@ def __init__(self,
...
-            if any(["model_state_epoch_" in x
-                    for x in os.listdir(self._serialization_dir)]):
-                logger.info("Loading model from checkpoint.")
-                epoch_counter = self._restore_checkpoint()
+        self._last_log = 0.0  # time of last logging
 
+        if serialization_dir is not None:
+            train_log = SummaryWriter(os.path.join(serialization_dir, "log", "train"))
+            validation_log = SummaryWriter(os.path.join(serialization_dir, "log", "validation"))
+            self._writer = TensorboardWriter(train_log, validation_log)
```
Comment (matt-gardner) [inline allennlp/training/trainer.py]: I think this name would be better if it had a clear reference to tensorboard in it.  E.g., I'd prefer `_tensorboard` to `_writer`.  `_writer` could be writing lots of different things, but `_tensorboard` is pretty obvious.

File: allennlp/training/trainer.py [hunk 2]
```diff
@@ -125,18 +143,16 @@ def __init__(self,
...
+        else:
+            self._writer = TensorboardWriter()
+
+    def _enable_gradient_clipping(self) -> None:
         if self._grad_clipping is not None:
             # Pylint is unable to tell that we're in the case that _glad_clipping is not None...
```
Comment (matt-gardner) [inline allennlp/training/trainer.py]: Just noticed this comment says "_glad_clipping" instead of "_grad_clipping"...  Care to fix my typo?

File: allennlp/training/trainer.py [hunk 3]
```diff
@@ -145,136 +161,243 @@ def train(self) -> None:
...
+        """
+        Performs gradient rescaling. Is a no-op if gradient rescaling is not enabled.
+        """
+        if self._grad_norm:
+            clip_grad_norm(self._model.parameters(), self._grad_norm)
+
+    def _compute_loss(self, batch: torch.Tensor, for_training: bool) -> torch.Tensor:
+        """
+        Does a forward pass on the given batch and returns the ``loss`` value in the result.
+        Also performs backpropagation if ``for_training`` is ``True``.
```
Comment (matt-gardner) [inline allennlp/training/trainer.py]: If this is going to perform backprop, I think it should say so in its name: `_compute_and_backprop_loss`.  It's pretty unintuitive to expect a `_compute_loss` function to also do backprop.

Comment (matt-gardner) [reply allennlp/training/trainer.py]: Actually, why put backprop in this function at all?  Why not just leave it in `_train_epoch`?  Seems cleaner.

File: allennlp/training/trainer.py [hunk 4]
```diff
@@ -145,136 +161,243 @@ def train(self) -> None:
...
+            # Advance the optimizer
+            self._optimizer.step()
+
+            # Update the description with the latest metrics
+            metrics = self._get_metrics_for_batch(train_loss, batch_num)
+            description = self._description_from_metrics(metrics)
+            train_generator_tqdm.set_description(description)
+
+            # Log parameter values to Tensorboard
+            batch_num_total = num_training_batches * epoch + batch_num
```
Comment (matt-gardner) [inline allennlp/training/trainer.py]: Can this block be pulled out into its own function, too?

File: allennlp/training/trainer.py [hunk 5]
```diff
@@ -145,136 +161,243 @@ def train(self) -> None:
...
+            self._optimizer.zero_grad()
+
+            # Compute loss for batch and backpropagate
+            loss = self._compute_loss(batch, for_training=True)
+
+            # Make sure Variable is on the cpu before converting to numpy.
+            # .cpu() is a no-op if you aren't using GPUs.
+            train_loss += loss.data.cpu().numpy()
+
+            # Rescale gradients
```
Comment (matt-gardner) [inline allennlp/training/trainer.py]: Most of these comments are just annotating a single line now, where the function being called already has a descriptive name.  I think you can remove these comments.

File: allennlp/training/trainer.py (hunk 4 above)
Comment (matt-gardner) [reply allennlp/training/trainer.py]: Also, doesn't look like all of the metrics are getting logged - shouldn't that be happening?

File: allennlp/training/trainer.py [hunk 6]
```diff
@@ -145,136 +161,243 @@ def train(self) -> None:
...
+                                                      batch_num_total)
+                self._writer.add_train_scalar("loss/loss_train", metrics["loss"], batch_num_total)
+
+            # Log progress in no-tqdm case
+            if self._no_tqdm and time.time() - self._last_log > self._log_interval:
+                logger.info("Batch %d/%d: %s", batch_num, num_training_batches, description)
+                self._last_log = time.time()
+
+        # Return the metrics (and reset them)
+        return self._get_metrics_for_batch(train_loss, batch_num, reset=True)
```
Comment (matt-gardner) [inline allennlp/training/trainer.py]: This is getting called twice, looks like.

File: allennlp/training/trainer.py [hunk 7]
```diff
@@ -145,136 +161,243 @@ def train(self) -> None:
...
+            loss = self._compute_loss(batch, for_training=False)
+            val_loss += loss.data.cpu().numpy()
+
+            # Update the description with the latest metrics
+            val_metrics = self._get_metrics_for_batch(val_loss, batch_num)
+            description = self._description_from_metrics(val_metrics)
+            val_generator_tqdm.set_description(description)
+
+            # Log progress in the no-tqdm case
+            if self._no_tqdm and time.time() - self._last_log > self._log_interval:
```
Comment (matt-gardner) [inline allennlp/training/trainer.py]: I think this was lost in `_train_epoch`.

Comment (matt-gardner) [reply allennlp/training/trainer.py]: You could take this block and the two lines above (`description = ...; ...tqdm.set_description()`) and put them into a function that gets called by both train and validation, too, I think.

File: allennlp/training/trainer.py [hunk 8]
```diff
@@ -145,136 +161,243 @@ def train(self) -> None:
...
+                logger.info("Batch %d/%d: %s", batch_num, num_validation_batches, description)
+                self._last_log = time.time()
+
+        # Log parameter values to Tensorboard
+        val_metrics = self._get_metrics_for_batch(val_loss, batch_num, reset=True)
+        message_template = "Training %s : %3f    Validation %s : %3f "
+        for name, value in metrics.items():
```
Comment (matt-gardner) [inline allennlp/training/trainer.py]: I'd _probably_ be fine with just having the `_train_epoch` function log the train metrics, and not have to pass them to this function.  If we do pass them here, I'd definitely at least call them `train_metrics` instead of just `metrics`.

File: allennlp/training/trainer.py [hunk 9]
```diff
@@ -145,136 +161,243 @@ def train(self) -> None:
...
+            logger.info(message_template, name, value, name, val_metrics[name])
+            self._writer.add_train_scalar(name, value, epoch)
+            self._writer.add_validation_scalar(name, val_metrics[name], epoch)
+
+        # Check validation metric for early stopping
+        this_epoch_val_metric = val_metrics[self._validation_metric]
```
Comment (matt-gardner) [inline allennlp/training/trainer.py]: Can this be its own function?

File: allennlp/training/trainer.py [hunk 10]
```diff
@@ -145,136 +161,243 @@ def train(self) -> None:
...
-                                  torch.optim.lr_scheduler.ReduceLROnPlateau):
-                        raise ConfigurationError("The reduce_on_plateau learning rate scheduler requires "
-                                                 "a validation metric to compute the schedule and therefore "
-                                                 "must be used with a validation dataset.")
-                    self._learning_rate_scheduler.step(epoch)
+            # Train one epoch and compute metrics.
+            metrics = self._train_epoch(epoch)
+
+            # Validate and move on to the next epoch
+            if self._should_stop(epoch, metrics, validation_metric_per_epoch):
```
Comment (matt-gardner) [inline allennlp/training/trainer.py]: I feel like this one went a bit too far - I'd probably put an explicit check for validation data here, and a call to `validate` if so, otherwise just save a checkpoint, then put the learning rate scheduler update stuff after that.  Having the `next_epoch{_no}_validation` seems a little bit much to me.

File: allennlp/training/trainer.py (hunk 1 above)
Comment (joelgrus) [reply allennlp/training/trainer.py]: done

File: allennlp/training/trainer.py (hunk 2 above)
Comment (joelgrus) [reply allennlp/training/trainer.py]: done

File: allennlp/training/trainer.py (hunk 3 above)
Comment (joelgrus) [reply allennlp/training/trainer.py]: done

File: allennlp/training/trainer.py (hunk 5 above)
Comment (joelgrus) [reply allennlp/training/trainer.py]: done

File: allennlp/training/trainer.py (hunk 6 above)
Comment (joelgrus) [reply allennlp/training/trainer.py]: are you sure? the previous call is per-batch inside the loop, this is at the end of the loop to reset the metrics. or am I missing one?

File: allennlp/training/trainer.py (hunk 4 above)
Comment (joelgrus) [reply allennlp/training/trainer.py]: I can add that

File: allennlp/training/trainer.py (hunk 6 above)
Comment (matt-gardner) [reply allennlp/training/trainer.py]: Ah, yeah, you're right.  I think the name referring to `for_batch` threw me off.  Not sure how to make the name better, though...

File: allennlp/training/trainer.py (hunk 9 above)
Comment (joelgrus) [reply allennlp/training/trainer.py]: done

File: allennlp/training/trainer.py (hunk 8 above)
Comment (joelgrus) [reply allennlp/training/trainer.py]: I changed the name

File: allennlp/training/trainer.py (hunk 6 above)
Comment (joelgrus) [reply allennlp/training/trainer.py]: I renamed to make this clearer

**Comment by joelgrus** [comment]
ok, I think this is ready for another look. I addressed most of the comments, there were a couple of "maybe factor this out" that I couldn't figure out a non-messy way to do, but I think it looks pretty good now.

File: allennlp/training/trainer.py [hunk 11]
```diff
@@ -31,6 +31,24 @@
...
+    """
+    def __init__(self, train_log: SummaryWriter = None, validation_log: SummaryWriter = None) -> None:
+        self._train_log = train_log
+        self._validation_log = validation_log
+
+    def add_train_scalar(self, name: str, value: float, global_step=None) -> None:
+        if self._train_log is not None:
+            self._train_log.add_scalar(name, value, global_step)
+
+    def add_validation_scalar(self, name: str, value: float, global_step=None) -> None:
```
Comment (DeNeutoy) [inline allennlp/training/trainer.py]: Why is global_step optional?

Comment (joelgrus) [reply allennlp/training/trainer.py]: I'm sure I must have had a good reason for it, but I couldn't tell you what it is now.

Comment (joelgrus) [reply allennlp/training/trainer.py]: I'll make it non-optional.
